lib/AST/Printer/Printer.cpp

```diff
@@ -640,4 +718,469 @@ PrintedToken PrintedTokenRange::operator[](size_t index) const {
+    }
+  }
+
+  return true;
+}
+
+bool PrintingPolicyAdaptor::ShouldPrintTemplate(
+    clang::ClassTemplatePartialSpecializationDecl *tpl) const {
+
+#ifndef PASTA_IN_BOOTSTRAP
+  if (pp) {
+    auto wrapped_tpl = DeclBuilder::Create<ClassTemplatePartialSpecializationDecl>(ast, tpl);
+    return pp->ShouldPrintTemplate(wrapped_tpl);
+  }
+#endif
+
+  if (!decl_to_print) {
+    return true;
+  }
+
+  if (tpl == decl_to_print) {
+    return true;
+  }
+
+  if (auto cspec = clang::dyn_cast<clang::ClassTemplateSpecializationDecl>(decl_to_print)) {
+    auto inst = cspec->getSpecializedTemplateOrPartial();
+    if (auto cpspec = inst.dyn_cast<clang::ClassTemplatePartialSpecializationDecl *>()) {
+      return cpspec->getCanonicalDecl() != tpl->getCanonicalDecl();
+    }
+  }
+
+  return true;
+}
+
+bool PrintingPolicyAdaptor::ShouldPrintTemplate(
+    clang::VarTemplatePartialSpecializationDecl *tpl) const {
+
+#ifndef PASTA_IN_BOOTSTRAP
+  if (pp) {
+    auto wrapped_tpl = DeclBuilder::Create<VarTemplatePartialSpecializationDecl>(ast, tpl);
+    return pp->ShouldPrintTemplate(wrapped_tpl);
+  }
+#endif
+
+  if (!decl_to_print) {
+    return true;
+  }
+
+  if (tpl == decl_to_print) {
+    return true;
+  }
+
+  if (auto cspec = clang::dyn_cast<clang::VarTemplateSpecializationDecl>(decl_to_print)) {
+    auto inst = cspec->getSpecializedTemplateOrPartial();
+    if (auto cpspec = inst.dyn_cast<clang::VarTemplatePartialSpecializationDecl *>()) {
+      return cpspec->getCanonicalDecl() != tpl->getCanonicalDecl();
+    }
+  }
+
+  return true;
+}
+
+bool PrintingPolicyAdaptor::ShouldPrintSpecialization(
+    clang::ClassTemplateDecl *tpl,
+    clang::ClassTemplateSpecializationDecl *spec) const {
+
+#ifndef PASTA_IN_BOOTSTRAP
+  if (pp) {
+    auto wrapped_tpl = DeclBuilder::Create<ClassTemplateDecl>(ast, tpl);
+    auto wrapped_spec = DeclBuilder::Create<ClassTemplateSpecializationDecl>(ast, spec);
+    return pp->ShouldPrintSpecialization(wrapped_tpl, wrapped_spec);
+  }
+#endif
+
+  return spec == decl_to_print;
+}
+
+bool PrintingPolicyAdaptor::ShouldPrintSpecialization(
+    clang::VarTemplateDecl *tpl,
+    clang::VarTemplateSpecializationDecl *spec) const {
+
+#ifndef PASTA_IN_BOOTSTRAP
+  if (pp) {
+    auto wrapped_tpl = DeclBuilder::Create<VarTemplateDecl>(ast, tpl);
+    auto wrapped_spec = DeclBuilder::Create<VarTemplateSpecializationDecl>(ast, spec);
+    return pp->ShouldPrintSpecialization(wrapped_tpl, wrapped_spec);
+  }
+#endif
+
+  return spec == decl_to_print;
+}
+
+bool PrintingPolicyAdaptor::ShouldPrintSpecialization(
+    clang::FunctionTemplateDecl *tpl, clang::FunctionDecl *spec) const {
+
+#ifndef PASTA_IN_BOOTSTRAP
+  if (pp) {
+    auto wrapped_tpl = DeclBuilder::Create<FunctionTemplateDecl>(ast, tpl);
+    auto wrapped_spec = DeclBuilder::Create<FunctionDecl>(ast, spec);
+    return pp->ShouldPrintSpecialization(wrapped_tpl, wrapped_spec);
```

Comment:
The template types are all different and need to be explicitly specified.

Reply:
```
template< typename A, typename B >
auto foo( ... )
{
...
}
```
But since you are always doing `tpl` and `spec` (and those are wrapped in the same way)
```
template< typename A, typename B >
auto wrap( B a )  { return DeclBuilder::Create< A >( ast, a ); }

return pp->ShouldPrintSpecialization( wrap< ... >( tpl ),
                                      wrap< ... >( spec ) );
```
I am not sure how much more it is worth to generalise (but I don't like the fallthrough of the bootstrap ifdef) but even with ^ it leads to nicer code because it streamlines the code.